### PR TITLE
Pattern & Sequence machines using nupic::Random

### LINKED ISSUE #2565: Pattern and Sequence machines shift to using nupic::Random
Currently the Pattern and Sequence machines in NuPIC use NumPy rand. This issue is to move the two machines across to use NuPIC Random.

### PR DESCRIPTION
**Pull request by rcrowder**
@chetan51 cc @oxtopus @scottpurdy Here's a PR showing the move to use nupic::Random in the Pattern and Sequence tests.

I may need some advice on how to squash some of these commits. Will try an interactive rebase next..

Fixes #2565

### CODE REVIEW

**Comment by oxtopus** [comment]
I'm guessing you want to keep "nupic::Random in Pattern and Sequence machines" and "Changing Pattern and Sequence machines to use UInt32 ndarrays" and squash the merge commits into one?

**Comment by rcrowder** [comment]
Yep. I tried a couple of rebases on another branch this weekend. The first seemed to go ok, but tricky with a further rebase. Although after GH didn't show what I was expecting, and the Windows GH GUI shows something else on the branch. The main problem faced was HEAD~n not working as expected.

File: nupic/data/generators/pattern_machine.py [hunk 1]
```diff
@@ -172,7 +173,7 @@ def _generate(self):
     """
     Generates set of random patterns.
     """
-    candidates = range(self._n)
+    candidates = np.array(range(self._n), np.uint32)
```
Comment (chetan51) [inline nupic/data/generators/pattern_machine.py]: Is this change necessary?

File: nupic/data/generators/sequence_machine.py [hunk 2]
```diff
@@ -132,13 +133,14 @@ def generateNumbers(self, numSequences, sequenceLength, sharedRange=None):
 
     for i in xrange(numSequences):
       start = i * sequenceLength
-      newNumbers = range(start, start + sequenceLength)
+      newNumbers = np.array(range(start, start + sequenceLength), np.uint32)
       self._random.shuffle(newNumbers)
 
       if sharedRange is not None:
+        newNumbers = list(newNumbers)
```
Comment (chetan51) [inline nupic/data/generators/sequence_machine.py]: Can move to above the conditional, and eliminate the change below.

**Comment by chetan51** [comment]
:+1:

**Comment by oxtopus** [comment]
I took a stab at rebasing it at https://github.com/numenta/nupic/compare/numenta:master...oxtopus:Pattern-Sequence-Machine-Random-Impl-rebased

I actually ended up cherry-picking the two commits in order after resetting to master.  You can close this PR in favor of #2566

**Comment by rcrowder** [comment]
Thanks for the rebase and explanation. The one change from Chetan is minor, should it be applied on my branch and picked up, or onto the new rebased one?

**Comment by oxtopus** [comment]
Apply to new.

**Comment by rcrowder** [comment]
Hmm, that didn't work :( I was able to clone your Nupic fork, checkout the rebased branch, make the change, commit, but then it errors during the push (remote: Permission to oxtopus/nupic.git denied to rcrowder). GH is also refusing to let me fork your nupic fork, I guess because I already have a fork of nupic? I guess last step is to create a PR from your fork?

**Comment by oxtopus** [comment]
Push to your fork rather than mine, then create a PR.  (I will close #2566)
